ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs

```diff
@@ -370,6 +370,11 @@ data TraceFetchClientState header =
      | AcknowledgedFetchRequest
          (FetchRequest header)
 
+       -- | Mark the point when fetch request for a fragment is actually sent
+       -- over the wire.
+     | SendFetchRequest
```

Comment:
Actually; I don't know what level of abstraction you're focused on, but I suppose you could get by via one of these tracers, matching on the `MsgRequestRange` constructor, instead of adding a constructor to `TraceFetchClientState`.

https://github.com/input-output-hk/ouroboros-network/blob/eb02c2d627b728a48449d314e56c347f1798503b/ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToNode.hs#L330-L331

I'm only pointing out what superficially seems to me like an option that doesn't involve changing the code.

Reply:
I would rather avoid enabling low-level protocol tracers, because they _seem_ to be more expensive to have enabled.

Maybe my intuition is off, though?

Reply:
These tracers are applied per message.

My anticipation is that a per-message tracer that does nothing for most messages `Tracer $ \case Foo -> ...; _ -> pure ()` wouldn't be distinguishable from a per-message tracer that does nothing for all messages (`nullTracer`). Specifically, I'd be surprised if enough inlining is happening for GHC to really generate significantly tighter code when the supplied tracer is `nullTracer`.

But, I haven't measured anything, so that's all just guess work.

Reply:
If a per-message tracer is doing something non-trivial for many messages, than that would seem likely to incur noticeable overhead.

Reply:
The problem with high-frequency tracers is that once you enable them, then all messages they emit are processed by the switchboard (which is the definition of "tracer enabledness" in the node), which is _slow_ -- and that's the primary reason why I tend to avoid them.

Reply:
Ah OK. I'm not really familiar with what you mean by switchboard, but I think I've seen something in the node code that makes the difference you're referring to.

👍